activation/poetdb.go

```diff
@@ -0,0 +1,150 @@
+package activation
+
+import (
+	"encoding/binary"
+	"fmt"
+	"github.com/spacemeshos/go-spacemesh/common"
+	"github.com/spacemeshos/go-spacemesh/database"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/types"
+	"github.com/spacemeshos/merkle-tree"
+	"github.com/spacemeshos/poet/hash"
+	"github.com/spacemeshos/poet/shared"
+	"github.com/spacemeshos/poet/verifier"
+	"github.com/spacemeshos/sha256-simd"
+)
+
+type PoetDb struct {
+	store                     database.Database
+	poetProofRefSubscriptions map[[32]byte][]chan []byte
+	log                       log.Log
+}
+
+func NewPoetDb(store database.Database, log log.Log) *PoetDb {
+	return &PoetDb{store: store, poetProofRefSubscriptions: make(map[[32]byte][]chan []byte), log: log}
+}
+
+func (db *PoetDb) ValidateAndStorePoetProof(proof types.PoetProof, poetId [types.PoetIdLength]byte, roundId uint64,
+	signature []byte) (processingIssue bool, err error) {
+
+	root, err := calcRoot(proof.Members)
+	if err != nil {
+		return true, fmt.Errorf("failed to calculate membership root for poetId %x round %d: %v",
+			poetId, roundId, err)
+	}
+	if err := validatePoet(root, proof.MerkleProof, proof.LeafCount); err != nil {
+		return false, fmt.Errorf("failed to validate poet proof for poetId %x round %d: %v",
+			poetId, roundId, err)
+	}
+
+	poetProof, err := types.InterfaceToBytes(&proof)
+	if err != nil {
+		return true, fmt.Errorf("failed to marshal poet proof for poetId %x round %d: %v",
+			poetId, roundId, err)
+	}
+
+	// TODO(noamnelke): validate signature (or extract public key and use for salting merkle hashes)
+
+	ref := sha256.Sum256(poetProof)
+
+	batch := db.store.NewBatch()
+	if err := batch.Put(ref[:], poetProof); err != nil {
+		return true, fmt.Errorf("failed to store poet proof for poetId %x round %d: %v",
+			poetId, roundId, err)
+	}
+	key := makeKey(poetId, roundId)
+	if err := batch.Put(key[:], ref[:]); err != nil {
+		return true, fmt.Errorf("failed to store poet proof index entry for poetId %x round %d: %v",
+			poetId, roundId, err)
+	}
+	if err := batch.Write(); err != nil {
+		return true, fmt.Errorf("failed to store poet proof and index for poetId %x round %d: %v",
+			poetId, roundId, err)
+	}
+	db.log.Info("stored proof for round %d PoET id %x", roundId, poetId)
+	db.publishPoetProofRef(key, ref[:])
```

Comment:
why do we publish the poet proof? isn't it already passed via gossip to everybody else?